taichi/program/snode_expr_utils.cpp

```diff
@@ -0,0 +1,110 @@
+#include "taichi/program/snode_expr_utils.h"
+
+namespace taichi {
+namespace lang {
+
+namespace {
+
+class GradInfoImpl final : public SNode::GradInfoProvider {
+ public:
+  explicit GradInfoImpl(GlobalVariableExpression *glb_var) : glb_var_(glb_var) {
+  }
+
+  bool is_primal() const override {
+    return glb_var_->is_primal;
+  }
+
+  SNode *grad_snode() const override {
+    auto &adj = glb_var_->adjoint;
+    if (adj.expr == nullptr) {
+      return nullptr;
+    }
+    return adj.snode();
+  }
+
+ private:
+  GlobalVariableExpression *glb_var_;
+};
+
+}  // namespace
+
+void place_child(Expr *expr_arg,
+                 const std::vector<int> &offset,
+                 SNode *parent,
+                 SNodeGlobalVarExprMap *snode_to_exprs) {
+  if (parent->type == SNodeType::root) {
+    // never directly place to root
+    auto &ds = parent->dense(std::vector<Index>(), {});
+    place_child(expr_arg, offset, &ds, snode_to_exprs);
+  } else {
+    TI_ASSERT(expr_arg->is<GlobalVariableExpression>());
+    auto glb_var_expr = expr_arg->cast<GlobalVariableExpression>();
+    TI_ERROR_IF(glb_var_expr->snode != nullptr,
+                "This variable has been placed.");
+    SNode *new_exp_snode = nullptr;
+    if (auto cft = glb_var_expr->dt->cast<CustomFloatType>()) {
```

Comment:
This part (around quant types) may be OK to be left inside `SNode` itself. But we can change that later if necessary..

Reply:
Sounds good. Let's decide after the quant API/impl materializes...